### DISCUSSION TITLE
Storing Uint8Arrays in a Supabase table

### DISCUSSION THREAD

**Discussion by cristiano-belloni** [supabase/supabase#2441] [Questions]
Hello, I'm trying to store incremental updates, generated by my frontend as `Uint8Array`s, in a Supabase table. My reasoning is that the best data type for byte arrays is `bytea`, so I generate my table like this:

```sql
CREATE TABLE incremental_updates (
  doc bigint NOT NULL,
  inserted_at timestamp without time zone DEFAULT timezone('utc'::text, now()) NOT NULL,
  data bytea NOT NULL,
  PRIMARY KEY(doc, inserted_at)
);
```

When I get an update, I do something like:

```ts
const { data, error } = await supabase
        .from("incremental_updates")
        .insert([{ data: update, doc: clientId }]);
      if (error) {
        console.error("Error!", error);
      } else if (data) {
        console.log("Inserted data into the server", data);
      }
```

The problem is that this example update, long 85 bytes:

`Uint8Array(85) [1, 1, 175, 138, 160, 164, 8, 1, 136, 175, 138, 160, 164, 8, 0, 1, 118, 3, 7, 99, 111, 110, 116, 101, 110, 116, 119, 17, 116, 104, 105, 115, 32, 105, 115, 32, 97, 110, 32, 117, 112, 100, 97, 116, 101, 9, 99, 114, 101, 97, 116, 101, 100, 65, 116, 123, 66, 119, 171, 229, 111, 180, 112, 0, 10, 109, 111, 100, 105, 102, 105, 101, 100, 65, 116, 123, 66, 119, 171, 229, 111, 180, 112, 0, 0]`

Seems to be stored on the database with this huge 1432 character record:

`\x7b2230223a312c2231223a312c2232223a3137352c2233223a3133382c2234223a3136302c2235223a3136342c2236223a382c2237223a312c2238223a3133362c2239223a3137352c223130223a3133382c223131223a3136302c223132223a3136342c223133223a382c223134223a302c223135223a312c223136223a3131382c223137223a332c223138223a372c223139223a39392c223230223a3131312c223231223a3131302c223232223a3131362c223233223a3130312c223234223a3131302c223235223a3131362c223236223a3131392c223237223a31372c223238223a3131362c223239223a3130342c223330223a3130352c223331223a3131352c223332223a33322c223333223a3130352c223334223a3131352c223335223a33322c223336223a39372c223337223a3131302c223338223a33322c223339223a3131372c223430223a3131322c223431223a3130302c223432223a39372c223433223a3131362c223434223a3130312c223435223a392c223436223a39392c223437223a3131342c223438223a3130312c223439223a39372c223530223a3131362c223531223a3130312c223532223a3130302c223533223a36352c223534223a3131362c223535223a3132332c223536223a36362c223537223a3131392c223538223a3137312c223539223a3232392c223630223a3131312c223631223a3138302c223632223a3131322c223633223a302c223634223a31302c223635223a3130392c223636223a3131312c223637223a3130302c223638223a3130352c223639223a3130322c223730223a3130352c223731223a3130312c223732223a3130302c223733223a36352c223734223a3131362c223735223a3132332c223736223a36362c223737223a3131392c223738223a3137312c223739223a3232392c223830223a3131312c223831223a3138302c223832223a3131322c223833223a302c223834223a307d`

What is the best, more efficient way to store small `Uint8Array` updates in my table?

**Comment by steve-chavez** [comment]
> Seems to be stored on the database with this huge 1432 character record:

This is postgres default [hex representation](https://www.postgresql.org/docs/current/datatype-binary.html#id-1.5.7.12.9) for bytea. If you encode it to "escape" you'll see the uints you sent:

```select
select encode('\x7b2230223a312c223122...', 'escape');

{"0":1,"1":1,"2":175,"3":138,"4":160,"5":164,"6":8,"7":1,"8":136...
```

**Reply by cristiano-belloni** [reply]
@steve-chavez do you mean I'm saving the `toString` output in my table, brackets, commas, quotes and all? That's not what I wanted 😅 

How can I save my Uint8Array binary *content* as an array of bytes? Like in the docs, if I have an array containing the bytes for `DEADBEEF`, having a corresponding `\xdeadbeef` in my table?

**Reply by canardos** [reply]
>_do you mean I'm saving the toString output in my table, brackets, commas, quotes and all?_

Yes. When you pass a JS number array to the Supabase JS API, the array is converted to a string, and each character is stored as a byte (the ASCII code) in the BYTEA column.

e.g. [1,2,3] becomes "[1,2,3]", which becomes {"type":"Buffer","data":[91,49,44,50,44,51,93]}

To store it correctly, you need to pass a string in the Postgres hex format.

e.g.:
```
const bytesToHex = (src: number[]) => "\\x" + src.reduce( (s,n) => s + n.toString(16).padStart(2, '0'), "");

supabase
    .from("incremental_updates")
    .insert([{ data: bytesToHex([1,2,3]), doc: 1 }]);

```

Similarly, when you retrieve the column, you'll need to decode on the client side.

**Reply by MahmoudMousaHamad** [reply]
How do you decode a byte array in hex?

**Comment by martinboksa** [comment]
Does anyone have a simple example, please? 🚀

**Comment by ashot** [comment]
You can the built-in utilities on the Buffer class to convert back and forth:

`Buffer.from(hex.toString('hex'), 'hex')`

you do still need to make the \x adjustment, so its `'\\x' + buffer.toString('hex')` and `Buffer.from(str.slice(2), 'hex')`

**Comment by markbulingit** [comment]
Learn from my mistake, saving Uint8List to column is just not possible, better to use storage feature
![image](https://github.com/supabase/supabase/assets/91706339/d4a407c6-fbc4-485a-a116-74615fb0b0a3)
